Extensions/Leaderboards/leaderboardstools.ts

```diff
@@ -45,74 +116,196 @@ namespace gdjs {
+            body: JSON.stringify({
+              playerName: formattedPlayerName,
+              score: score,
+              sessionId: sessionId,
+              clientPlayerId: playerId,
+              location: location,
+            }),
+            method: 'POST',
+            headers: { 'Content-Type': 'application/json' },
+          }
+        ).then(
+          (response) => {
+            if (!response.ok) {
+              const errorCode = response.status.toString();
+              logger.error(
+                'Server responded with an error:',
+                errorCode,
+                response.statusText
+              );
+              scoreSavingData.setError(errorCode);
+              errorVar.setString(errorCode);
+              return;
             }
-          )
-            .then((response) => {
-              _scoreLastSentAt = Date.now();
-              if (!response.ok) {
-                errorVar.setString(response.status.toString());
-                _lastErrorCode = response.status;
-                return response.statusText;
-              } else {
-                _lastScore = score;
-                _lastPlayerName = playerName;
-                _lastErrorCode = response.status;
-                return response.text();
+
+            scoreSavingData.closeSaving();
+
+            response.text().then(
+              (text) => {
+                responseVar.setString(text);
+              },
+              (error) => {
+                logger.warn(
+                  'An error occurred when reading response but score has been saved:',
+                  error
+                );
+                responseVar.setString('CANNOT_READ_RESPONSE');
               }
-            })
-            .then((data) => {
-              responseVar.setString(data);
-            })
-            .catch((error) => {
-              logger.warn(
-                `Error while submitting a leaderboard score: ${error}`
-              );
-              errorVar.setString('REQUEST_NOT_SENT');
-              _lastErrorCode = 400;
-            });
+            );
+          },
+          (error) => {
+            logger.error('Error while submitting a leaderboard score:', error);
+            const errorCode = 'REQUEST_NOT_SENT';
+            scoreSavingData.setError(errorCode);
+            errorVar.setString(errorCode);
+          }
+        );
+      };
+
+      const getLeaderboardOrLastLeaderboardState = function (
+        state: string,
+        leaderboardId?: string
+      ): boolean {
+        if (leaderboardId) {
+          if (!_savingDataByLeaderboard[leaderboardId]) {
+            return false;
+          }
+          return _savingDataByLeaderboard[leaderboardId][state];
         }
+        const lastScoreSavingData = getLastScoreSavingData({
+          hasEnded: state === 'hasScoreBeenSaved',
+        });
+        if (!lastScoreSavingData) {
+          return false;
+        }
+        return lastScoreSavingData[state];
```

Comment:
Nits:
- Prefer to return the ScoreSavingData or null
- Choose a name (ScoreSavingData or ScoreSavingState) and use this rather than "LeaderboardState" to not create multiple concepts in the head of the reader.

```suggestion
      const getLeaderboardOrLastLeaderboardSavingData = function (
        leaderboardId?: string
      ): boolean {
        if (leaderboardId) {
          return _savingDataByLeaderboard[leaderboardId] || null;
        }
        const lastScoreSavingData = getLastScoreSavingData({
          hasEnded: state === 'hasScoreBeenSaved',
        });
        return lastScoreSavingData || null;
```